src/Nethermind/Nethermind.Trie/Pruning/OverlayTrieStore.cs

```diff
@@ -1,38 +1,44 @@
 // SPDX-FileCopyrightText: 2024 Demerzel Solutions Limited
 // SPDX-License-Identifier: LGPL-3.0-only
 
-using System;
 using Nethermind.Core;
 using Nethermind.Core.Crypto;
-using Nethermind.Logging;
 
 namespace Nethermind.Trie.Pruning;
 
-public class OverlayTrieStore(IKeyValueStoreWithBatching? keyValueStore, IReadOnlyTrieStore store, ILogManager? logManager) : TrieStore(keyValueStore, logManager)
+public class OverlayTrieStore(IKeyValueStoreWithBatching keyValueStore, IReadOnlyTrieStore store) : ITrieStore
 {
-    public override bool IsPersisted(Hash256? address, in TreePath path, in ValueHash256 keccak) =>
-        base.IsPersisted(address, in path, in keccak) || store.IsPersisted(address, in path, in keccak);
+    private readonly INodeStorage _nodeStorage = new NodeStorage(keyValueStore);
 
-    internal override TrieNode FindCachedOrUnknown(Hash256? address, in TreePath path, Hash256? hash, bool isReadOnly)
+    public void Dispose()
     {
-        ArgumentNullException.ThrowIfNull(hash);
-
-        TrieNode node = base.FindCachedOrUnknown(address, in path, hash, isReadOnly);
-        return node.NodeType == NodeType.Unknown
-            ? store.FindCachedOrUnknown(address, in path, hash) // no need to pass isReadOnly - IReadOnlyTrieStore overrides it as true
-            : node;
+        store.Dispose();
     }
 
-    public override byte[]? LoadRlp(Hash256? address, in TreePath path, Hash256 hash, ReadFlags flags = ReadFlags.None) =>
-        base.TryLoadRlp(address, in path, hash, flags) ?? store.LoadRlp(address, in path, hash, flags);
-
-    public override byte[]? TryLoadRlp(Hash256? address, in TreePath path, Hash256 hash, ReadFlags flags = ReadFlags.None) =>
-        base.TryLoadRlp(address, in path, hash, flags) ?? store.TryLoadRlp(address, in path, hash, flags);
+    public TrieNode FindCachedOrUnknown(Hash256? address, in TreePath path, Hash256 hash)
+    {
+        return store.FindCachedOrUnknown(address, in path, hash);
+    }
 
-    protected override void VerifyNewCommitSet(long blockNumber)
+    public byte[]? LoadRlp(Hash256? address, in TreePath path, Hash256 hash, ReadFlags flags = ReadFlags.None)
     {
-        // Skip checks, as override can be applied using the same block number or without a state root
+        byte[]? rlp = TryLoadRlp(address, in path, hash, flags);
+        if (rlp is null) throw new MissingTrieNodeException("Missing RLP node", address, path, hash);
+        return rlp;
     }
 
-    public void ResetOverrides() => ClearCache();
+    public byte[]? TryLoadRlp(Hash256? address, in TreePath path, Hash256 hash, ReadFlags flags = ReadFlags.None) => store.TryLoadRlp(address, in path, hash, flags) ?? _nodeStorage.Get(address, in path, hash, flags);
```

Comment:
I think you have order wrong, `_nodeStorage` which is based on `keyValueStore` should be preferenced, as it is where the overrides live. Similar in all other methods